### PR TITLE
Secure upgrade HLD

### PR DESCRIPTION
**Pull request by ycoheNvidia**
This PR contains the Secure Upgrade HLD.

| PR title | state | context |
| ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------ |
| [[sonic-buildimage] Added support for Secure Upgrade](https://github.com/Azure/sonic-buildimage/pull/11862) | GitHub issue/pull request detail | GitHub pull request check contexts |
| [[sonic-utilities] Added support for Secure Upgrade](https://github.com/Azure/sonic-utilities/pull/2337) | GitHub issue/pull request detail | GitHub pull request check contexts |

### CODE REVIEW

**Comment by zhangyanzhao** [comment]
@madhupalu from Aviz Networks will help to review this.

**Comment by liat-grozovik** [comment]
@yxieca , @madhupalu as stated in the 202211 release features list you are the reviewers. Could you please help to review?
Nvidia will soon have the code PRs avaialble thus need to close the HLD first.

**Comment by ycoheNvidia** [comment]
| PR title | state | context |
| ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------ |
| [[sonic-buildimage] Added support for Secure Upgrade]([https://github.com/Azure/sonic-buildimage/pull/11862](https://github.com/Azure/sonic-buildimage/pull/10323)) | GitHub issue/pull request detail) | GitHub pull request check contexts) |
| [[sonic-utilities Added support for Secure Upgrade](https://github.com/Azure/sonic-utilities/pull/2337) | GitHub issue/pull request detail) | GitHub pull request check contexts |

**Comment by liat-grozovik** [comment]
@yxieca , @madhupalu kindly reminder for review comments.

**Comment by yxieca** [comment]
@vaibhavhd  to check from upgrade perspective. @xumia to check from secure boot/upgrade perspective.

File: doc/secure_upgrade/secure_upgrade.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	CMS - Cryptographic Message Syntax
+
+###  1.4. <a name='Overview'></a>Overview
+
+Secure installation and upgrade of OS are setting a requirement to ensure the image was not modified since it was built by the vendor. This is a basic requirement as part of general OS security. In order to help the credibility of the OS image, secure upgrade mechanism guarantees the image was not modified by any unauthorized entity since its original creation.
+
+###  1.5. <a name='Requirements'></a>Requirements
+
+We want to enable secure upgrade of SONIC. This will include secure upgrade of SONIC while installing SONIC via ONIE or SONIC. The feature requires a signing process which will include a private key and certificate. The verification process on the other end requires a public key certificate for validation. Both processes are required to be supported by various scenarios such as development and release.
+In order to do secure installation of SONIC from ONIE, it has to include the flowing secure upgrade verification commit TBD (on pull request - waiting for it to be created)
```
Comment (madhupalu) [inline doc/secure_upgrade/secure_upgrade.md:65]: correct typo "flowing" here - is the secure upgrade verification commit ready? if yes please share it

File: doc/secure_upgrade/secure_upgrade.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+![signing_options](signing_options.png)
+
+####  1.7.2. <a name='Verificationprocess'></a>Verification process
+- SONIC install – changes will be mainly done in sonic-installer main.py where bootloader methods are being called and will only affect 'sonic-installer install' command (other sonic-installer commands will not be affected). The installation process is triggered by install.sh script, which is different for every platform.
+In our opinion, python modifications will be easier to add and maintain than install.sh changes, as changing install.sh (for each platform) will require much more sources to maintain and test (different HW to check on, test cases for each HW, etc.).
+- A signed image will go through a verification process before installation, as main.py will call a dedicated bash script to split the signature from the image and check it with a public key certificate.
+Verification script called verify_image_sign.sh will verify using OpenSSL cms. Verification is enabled only if efi tools are enabled, and "Secure Boot" flag is enabled in BOIS. Accordingly, the certificate will be fetched from BIOS using EFI tools.
```
Comment (madhupalu) [inline doc/secure_upgrade/secure_upgrade.md:102]: correct BOIS type

File: doc/secure_upgrade/secure_upgrade.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+####  1.9.2. <a name='ConfigDBEnhancements'></a>Config DB Enhancements  
+
+NA
+		
+###  1.10. <a name='WarmbootandFastbootDesignImpact'></a>Warmboot and Fastboot Design Impact  
+NA
+
+###  1.11. <a name='RestrictionsLimitations'></a> Restrictions/Limitations  
+####  1.11.1. <a name='uefitoolsandsecurebootenabled'></a> uefi tools and secure boot enabled
+Device arch should support secure boot including UEFI tools.
```
Comment (madhupalu) [inline doc/secure_upgrade/secure_upgrade.md:130]: what are the UEFI tools being used to support secure upgrade process?

File: doc/secure_upgrade/secure_upgrade.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+This feature is dependent on BIOS secure boot flag being enabled.
+####  1.11.2. <a name='ONIEsecureupgradeverificationcode'></a> ONIE secure upgrade verification code
+Secure installation of SONIC from ONIE requires a specific verification code, which can be picked from the following ONIE pull request TBD (waiting for it to be created)
```
Comment (madhupalu) [inline doc/secure_upgrade/secure_upgrade.md:133]: Is the ONIE pull request is ready? please share if it is

File: doc/secure_upgrade/secure_upgrade.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+###  1.7. <a name='High-LevelDesign'></a>High-Level Design
+
+
+This feature has 2 flows to be supported:
+	
+####  1.7.1. <a name='Signingprocess'></a>Signing process  
+
+Sign sonic image during build process in build_image.sh script, right after generation.
+We want to add the flag SECURE_UPGRADE_MODE to allow several modes of secure upgrade. When SECURE_UPGRADE_MODE == 'no_sign' no change will be made to the current build process. In case of SECURE_UPGRADE_MODE == 'dev', image will be signed by development script and when SECURE_UPGRADE_MODE == 'prod' - image will be signed by production script. In addition, we want to add the flags SECURE_UPGRADE_DEV_SIGNING_KEY and SECURE_UPGRADE_DEV_SIGNING_CERT, to provide the option to give path for development keys during the build.
```
Comment (madhupalu) [inline doc/secure_upgrade/secure_upgrade.md:86]: How is the signing prod process is different from dev? How do you secure production keys? do you use vault kind of infra?

File: doc/secure_upgrade/secure_upgrade.md (hunk 1 above)
Comment (davidpil2002) [reply doc/secure_upgrade/secure_upgrade.md:65]: fix the typo, and for the moment the "commit" is waiting for review, added the current PR in the HLD

File: doc/secure_upgrade/secure_upgrade.md (hunk 2 above)
Comment (davidpil2002) [reply doc/secure_upgrade/secure_upgrade.md:102]: done

File: doc/secure_upgrade/secure_upgrade.md (hunk 3 above)
Comment (davidpil2002) [reply doc/secure_upgrade/secure_upgrade.md:130]: added the list and url

File: doc/secure_upgrade/secure_upgrade.md (hunk 4 above)
Comment (davidpil2002) [reply doc/secure_upgrade/secure_upgrade.md:133]: still not

File: doc/secure_upgrade/secure_upgrade.md (hunk 5 above)
Comment (davidpil2002) [reply doc/secure_upgrade/secure_upgrade.md:86]: The feature is including the code that supports secure upgrade using development keys. And any Vendor should decide his own design flow for production. 
In general, in the build flow when the user is building the image with "prod" flag (as explained in chapter 1.7.1. Signing process) the build flow will call a signing script that should be coded by the vendor, this script should send the image to a different server "sign server" and this server should return the image signed.

**Comment by Yarden-Z** [comment]
@yxieca , @xumia can you please approve this?
We fixed the latest fault we had here and answered all the review questions.

**Comment by liat-grozovik** [comment]
This is the last call for the HLD review. Kindly reminder for those who provided feedback or still need to provide to have them available this week
@zhangyanzhao as this is for 202211 we need to close the loop. Also code PR is available for review

**Comment by Yarden-Z** [comment]
@xumia / @zhangyanzhao Is there any update on this HLD review?

**Comment by zhangyanzhao** [comment]
@yxieca  @madhupalu can you please help to review this HLD and related code PRs? Need your approval after your review. Thanks.

File: doc/secure_upgrade/secure_upgrade.md (hunk 5 above)
Comment (xumia) [reply doc/secure_upgrade/secure_upgrade.md:86]: @davidpil2002 , we have a different environment for build production and dev, production is always built in security environment. Is it possible to use the same process for both production and dev? The only different is the certificates used, but the build steps are all the same.
It is good to provide an additional script to sign images from the "sign server", but can we support to the simple scenario to support "sign server" and build on the same machine? How about removing the "_DEV_" in the parameters?

Comment (ycoheNvidia) [reply doc/secure_upgrade/secure_upgrade.md:86]: @xumia thank you for the question. In many cases, different customers have different production servers with different environments. Therefor no single script will be applicable here. In the case you suggest - it is possible to copy the dev key as an argument to prod script, but this is a specific use case

**Review by xumia — APPROVED** [review]

**Comment by liat-grozovik** [comment]
The feature and its current content is expected to be in 202305 and into 202211 as originally aimed.

**Pull request merged**